WordPress/Classes/ViewRelated/Domains/View Models/SiteDomainsViewModel.swift

```diff
@@ -1,112 +1,156 @@
+                wpcomDomains.append(domain)
+            } else {
+                otherDomains.append(domain)
+            }
+        }
+
+        return Self.buildFreeDomainSections(from: blog, wpComDomains: wpcomDomains) + Self.buildDomainsSections(from: blog, domains: otherDomains)
     }
 
-    private static func buildFreeDomainSections(from blog: Blog) -> [Section] {
-        guard let freeDomain = blog.freeDomain else { return [] }
-        return [Section(
-            title: Strings.freeDomainSectionTitle,
-            footer: blog.freeDomainIsPrimary ? Strings.primaryDomainDescription : nil,
-            content: .rows([.init(
-                name: blog.freeSiteAddress,
-                description: nil,
-                status: nil,
-                expiryDate: DomainExpiryDateFormatter.expiryDate(for: freeDomain),
-                isPrimary: freeDomain.isPrimaryDomain
-            )])
-        )]
+    private static func buildFreeDomainSections(from blog: Blog, wpComDomains: [DomainsService.AllDomainsListItem]) -> [Section] {
+        let blogWpComDomains = wpComDomains.filter { $0.blogId == blog.dotComID?.intValue }
+        guard let freeDomain = blogWpComDomains.count > 1 ? blogWpComDomains.first(where: { $0.isWpcomStagingDomain }) : blogWpComDomains.first else {
+            return []
+        }
+
+        return [
+            Section(
+                title: Strings.freeDomainSectionTitle,
+                footer: blog.freeDomainIsPrimary ? Strings.primaryDomainDescription : nil,
+                content: .rows([.init(
+                    viewModel: .init(
+                        name: freeDomain.domain,
+                        description: nil,
+                        status: nil,
+                        expiryDate: AllDomainsListItemViewModel.expiryDate(from: freeDomain),
+                        isPrimary: blog.freeDomainIsPrimary
+                    ),
+                    navigation: nil)])
+            )
+        ]
     }
 
-    private static func buildDomainsSections(from blog: Blog) -> [Section] {
+    private static func buildDomainsSections(from blog: Blog, domains: [DomainsService.AllDomainsListItem]) -> [Section] {
         var sections: [Section] = []
 
-        let primaryDomain = blog.domainsList.first(where: { $0.domain.isPrimaryDomain })
-        let otherDomains = blog.domainsList.filter { !$0.domain.isPrimaryDomain }
+        let primaryDomainName = blog.domainsList.first(where: { $0.domain.isPrimaryDomain })?.domain.domainName
+        var primaryDomain: DomainsService.AllDomainsListItem?
+        var otherDomains: [DomainsService.AllDomainsListItem] = []
+
+        for domain in domains {
+            if domain.blogId == blog.dotComID?.intValue {
+                if primaryDomainName == domain.domain {
+                    primaryDomain = domain
+                } else {
+                    otherDomains.append(domain)
+                }
+            }
+        }
```

Comment:
nitpick: This alternative could also work:
```suggestion
        let blogDomains = domains.filter({ $0.blogId == blog.dotComID?.intValue })
        let primaryDomain = blogDomains.first(where: { primaryDomainName == $0.domain })
        let otherDomains = blogDomains.filter({ primaryDomainName != $0.domain })
```

Reply:
I did all this to have fewer iterations but that's probably unneeded optimization 😀. 2N or 3N is still N. There're no nested loops here. I will use your proposal 👍